### ISSUE TITLE
[SPIKE] Research on OpenLineage Integration

### ISSUE THREAD

**Issue by kaxil** [astronomer/astronomer-providers#204]
- We want to start an initial investigation on how we can emit meta to Datakin/OpenLinage
- We will take `BigQueryInsertJobOperatorAsync` as our first operator for this and see how we can retrieve things like "data processed", "rows affected" and publish that to OpenLineage.

OpenLineage spec: [link](https://docs.google.com/document/d/1vPsvHejQ24xTbzpz_LYSf0_ixk9oUuBiEUHVEaF9J2U/edit?usp=sharing)

Tasks:

- [x] Understand basic concepts of lineage, watch Airflow OpenLineage integration videos
- [x] Try a sample DAG to see lineage getting collected in Datakin
- [x] Explore and understand BigQuery
- [x] Understand how to write extractors for OpenLineage
- [x] Explore what metrics can be exported for BigQueryInsertJobOperator
- [x] Write extractors for the BigQueryInsertJobOperator

**Comment by pankajkoti** [comment]
Started understanding the concepts of lineage, exploring OpenLineage documentation. Next step is to watch a couple of videos for Airflow OpenLineage integration and try out a sample DAG and observe lineage in Datakin.

**Comment by kaxil** [comment]
I watched the following video too which was very helpful in understanding the concepts:

- https://www.youtube.com/watch?v=LRr-ja8_Wjs

The following links were also helpful:

- https://github.com/OpenLineage/OpenLineage/tree/main/integration/airflow
- https://github.com/OpenLineage/OpenLineage/pull/508 (This will use Listener Plugin from the upcoming Airflow release 2.3 that was added in https://github.com/apache/airflow/pull/20443)

**Comment by pankajkoti** [comment]
Was able to write a custom extractor for the `BigQueryInsertJobOperatorAsync` operator. However, the common util implementation from OpenLineage is not dumping statistics as expected. Will sync up with Maciej on Tuesday once he is back from vacation to check what is happening wrong. Although I believe this is an issue on OpenLineage implementation, I will try to debug and confirm this.

**Comment by pankajkoti** [comment]
The DAG runs fine now, but still the extractor is not extracting relevant metrics as expected. We are getting the below error from OpenLineage implementation and Maciej Obuchowski from GetInData (Datakin's consultancy agency) said he will create an issue for it and work on fixing it:

```
[2022-04-19, 11:07:33 UTC] {_metadata.py:99} WARNING - Compute Engine Metadata server unavailable on attempt 2 of 3. Reason: [Errno 111] Connection refused
[2022-04-19, 11:07:33 UTC] {_metadata.py:99} WARNING - Compute Engine Metadata server unavailable on attempt 3 of 3. Reason: [Errno 111] Connection refused
[2022-04-19, 11:07:33 UTC] {_default.py:286} WARNING - Authentication failed using Compute Engine authentication due to unavailable metadata server.
[2022-04-19, 11:07:33 UTC] {__init__.py:93} ERROR - Failed to extract metadata Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started task_type=BigQueryInsertJobOperatorAsync ***_dag_id=example_async_bigquery_queries task_id=insert_query_job ***_run_id=manual__2022-04-19T11:05:56.220393+00:00 
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/openlineage/lineage_backend/__init__.py", line 85, in _extract_metadata
    task_metadata = self._extract(extractor, task_instance)
  File "/usr/local/lib/python3.9/site-packages/openlineage/lineage_backend/__init__.py", line 106, in _extract
    task_metadata = extractor.extract_on_complete(task_instance)
  File "/usr/local/airflow/astronomer_providers/astronomer/providers/google/cloud/extractors/bigquery_extractor.py", line 63, in extract_on_complete
    stats = BigQueryDatasetsProvider().get_facets(bigquery_job_id)
  File "/usr/local/lib/python3.9/site-packages/openlineage/common/provider/bigquery.py", line 91, in __init__
    self.client = Client()
  File "/usr/local/lib/python3.9/site-packages/google/cloud/bigquery/client.py", line 231, in __init__
    super(Client, self).__init__(
  File "/usr/local/lib/python3.9/site-packages/google/cloud/client/__init__.py", line 320, in __init__
    _ClientProjectMixin.__init__(self, project=project, credentials=credentials)
  File "/usr/local/lib/python3.9/site-packages/google/cloud/client/__init__.py", line 268, in __init__
    project = self._determine_default(project)
  File "/usr/local/lib/python3.9/site-packages/google/cloud/client/__init__.py", line 287, in _determine_default
    return _determine_default_project(project)
  File "/usr/local/lib/python3.9/site-packages/google/cloud/_helpers/__init__.py", line 152, in _determine_default_project
    _, project = google.auth.default()
  File "/usr/local/lib/python3.9/site-packages/google/auth/_default.py", line 488, in default
    raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started
[2022-04-19, 11:07:34 UTC] {client.py:74} INFO - Constructing openlineage client to send events to http://host.docker.internal:5050/
```

**Comment by pankajkoti** [comment]
The activity is blocked on issue: https://github.com/OpenLineage/OpenLineage/issues/649 and has an open PR being worked upon by Maciej https://github.com/OpenLineage/OpenLineage/pull/691

The action items post this are:
1. Test E2E with the changes in the above PRs
2. Once E2E test works with no additional issues, address review comments on https://github.com/astronomer/astronomer-providers/pull/232 and close the ticket.

**Comment by pankajkoti** [comment]
The issue is resolved and I am able to get expected extractor event after introducing Maciej's change in my POC PR.

A sample event looks like the following:
`{
    "eventTime": "2022-04-29T05:23:41.450000Z",
    "eventType": "COMPLETE",
    "inputs": [
        {
            "facets": {
                "dataSource": {
                    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.2/integration/airflow",
                    "_schemaURL": "https://raw.githubusercontent.com/OpenLineage/OpenLineage/main/spec/OpenLineage.json#/definitions/DataSourceDatasetFacet",
                    "name": "bigquery",
                    "uri": "bigquery"
                },
                "schema": {
                    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.2/integration/airflow",
                    "_schemaURL": "https://raw.githubusercontent.com/OpenLineage/OpenLineage/main/spec/OpenLineage.json#/definitions/SchemaDatasetFacet",
                    "fields": [
                        {
                            "name": "value",
                            "type": "INTEGER"
                        },
                        {
                            "name": "name",
                            "type": "STRING"
                        },
                        {
                            "name": "ds",
                            "type": "DATE"
                        }
                    ]
                }
            },
            "name": "astronomer-airflow-providers.astro_dataset.table1",
            "namespace": "bigquery"
        }
    ],
    "job": {
        "facets": {},
        "name": "example_async_bigquery_queries.insert_query_job",
        "namespace": "dev"
    },
    "outputs": [
        {
            "facets": {
                "dataSource": {
                    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.2/integration/airflow",
                    "_schemaURL": "https://raw.githubusercontent.com/OpenLineage/OpenLineage/main/spec/OpenLineage.json#/definitions/DataSourceDatasetFacet",
                    "name": "bigquery",
                    "uri": "bigquery"
                },
                "schema": {
                    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.2/integration/airflow",
                    "_schemaURL": "https://raw.githubusercontent.com/OpenLineage/OpenLineage/main/spec/OpenLineage.json#/definitions/SchemaDatasetFacet",
                    "fields": [
                        {
                            "name": "value",
                            "type": "INTEGER"
                        },
                        {
                            "name": "name",
                            "type": "STRING"
                        },
                        {
                            "name": "ds",
                            "type": "DATE"
                        }
                    ]
                },
                "stats": {
                    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.2/integration/airflow",
                    "_schemaURL": "https://github.com/OpenLineage/OpenLineage/tree/main/integration/common/openlineage/schema/bq-statistics-dataset-facet.json",
                    "rowCount": 2,
                    "size": 0
                }
            },
            "name": "astronomer-airflow-providers.astro_dataset.table1",
            "namespace": "bigquery",
            "outputFacets": {
                "outputStatistics": {
                    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.2/integration/airflow",
                    "_schemaURL": "https://raw.githubusercontent.com/OpenLineage/OpenLineage/main/spec/OpenLineage.json#/definitions/OutputStatisticsOutputDatasetFacet",
                    "rowCount": 2,
                    "size": 0
                }
            }
        }
    ],
    "producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.2/integration/airflow",
    "run": {
        "facets": {
            "bigQuery_job": {
                "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.2/integration/airflow",
                "_schemaURL": "https://github.com/OpenLineage/OpenLineage/tree/main/integration/common/openlineage/schema/bq-statistics-run-facet.json",
                "billedBytes": 0,
                "cached": false,
                "properties": "{\"kind\": \"bigquery#job\", \"etag\": \"O4Oasb14KhLJ9NzTEhJM2g==\", \"id\": \"astronomer-airflow-providers:US.airflow_example_async_bigquery_queries_insert_query_job_2022_04_29T05_23_10_598413_00_00_22e44b8586a4974727d996d2f1447b33\", \"selfLink\": \"https://bigquery.googleapis.com/bigquery/v2/projects/astronomer-airflow-providers/jobs/airflow_example_async_bigquery_queries_insert_query_job_2022_04_29T05_23_10_598413_00_00_22e44b8586a4974727d996d2f1447b33?location=US\", \"user_email\": \"pankaj-koti-svc-account@astronomer-airflow-providers.iam.gserviceaccount.com\", \"jobReference\": {\"projectId\": \"astronomer-airflow-providers\", \"jobId\": \"airflow_example_async_bigquery_queries_insert_query_job_2022_04_29T05_23_10_598413_00_00_22e44b8586a4974727d996d2f1447b33\", \"location\": \"US\"}, \"statistics\": {\"creationTime\": 1651209809635.0, \"startTime\": 1651209809751.0, \"endTime\": 1651209811000.0, \"totalBytesProcessed\": \"0\", \"query\": {\"queryPlan\": [{\"name\": \"S00: Input\", \"id\": \"0\", \"startMs\": \"1651209809818\", \"endMs\": \"1651209809825\", \"waitRatioAvg\": 0.018404907975460124, \"waitMsAvg\": \"3\", \"waitRatioMax\": 0.018404907975460124, \"waitMsMax\": \"3\", \"readRatioAvg\": 0, \"readMsAvg\": \"0\", \"readRatioMax\": 0, \"readMsMax\": \"0\", \"computeRatioAvg\": 0.018404907975460124, \"computeMsAvg\": \"3\", \"computeRatioMax\": 0.018404907975460124, \"computeMsMax\": \"3\", \"writeRatioAvg\": 0.018404907975460124, \"writeMsAvg\": \"3\", \"writeRatioMax\": 0.018404907975460124, \"writeMsMax\": \"3\", \"shuffleOutputBytes\": \"57\", \"shuffleOutputBytesSpilled\": \"0\", \"recordsRead\": \"0\", \"recordsWritten\": \"2\", \"parallelInputs\": \"1\", \"completedParallelInputs\": \"1\", \"status\": \"COMPLETE\", \"steps\": [{\"kind\": \"COMPUTE\", \"substeps\": [\"$10 := 42\", \"$11 := 'monthy python'\", \"$12 := 19111\"]}, {\"kind\": \"COMPUTE\", \"substeps\": [\"$20 := 42\", \"$21 := 'fishy fish'\", \"$22 := 19111\"]}, {\"kind\": \"WRITE\", \"substeps\": [\"$1, $2, $3\", \"TO __stage00_output\"]}], \"slotMs\": \"12\"}, {\"name\": \"S01: Coalesce\", \"id\": \"1\", \"startMs\": \"1651209809859\", \"endMs\": \"1651209809932\", \"inputStages\": [\"0\"], \"waitRatioAvg\": 0.1411042944785276, \"waitMsAvg\": \"23\", \"waitRatioMax\": 0.147239263803681, \"waitMsMax\": \"24\", \"readRatioAvg\": 0, \"readMsAvg\": \"0\", \"readRatioMax\": 0, \"readMsMax\": \"0\", \"computeRatioAvg\": 0.03067484662576687, \"computeMsAvg\": \"5\", \"computeRatioMax\": 0.04294478527607362, \"computeMsMax\": \"7\", \"writeRatioAvg\": 0.1411042944785276, \"writeMsAvg\": \"23\", \"writeRatioMax\": 0.18404907975460122, \"writeMsMax\": \"30\", \"shuffleOutputBytes\": \"57\", \"shuffleOutputBytesSpilled\": \"0\", \"recordsRead\": \"2\", \"recordsWritten\": \"2\", \"parallelInputs\": \"50\", \"completedParallelInputs\": \"50\", \"status\": \"COMPLETE\", \"steps\": [{\"kind\": \"READ\", \"substeps\": [\"FROM __stage00_output\"]}], \"slotMs\": \"4212\"}, {\"name\": \"S02: Output\", \"id\": \"2\", \"startMs\": \"1651209810000\", \"endMs\": \"1651209810148\", \"inputStages\": [\"1\"], \"waitRatioAvg\": 1, \"waitMsAvg\": \"163\", \"waitRatioMax\": 1, \"waitMsMax\": \"163\", \"readRatioAvg\": 0, \"readMsAvg\": \"0\", \"readRatioMax\": 0, \"readMsMax\": \"0\", \"computeRatioAvg\": 0.03680981595092025, \"computeMsAvg\": \"6\", \"computeRatioMax\": 0.03680981595092025, \"computeMsMax\": \"6\", \"writeRatioAvg\": 0.8466257668711656, \"writeMsAvg\": \"138\", \"writeRatioMax\": 0.8466257668711656, \"writeMsMax\": \"138\", \"shuffleOutputBytes\": \"0\", \"shuffleOutputBytesSpilled\": \"0\", \"recordsRead\": \"2\", \"recordsWritten\": \"2\", \"parallelInputs\": \"1\", \"completedParallelInputs\": \"1\", \"status\": \"COMPLETE\", \"steps\": [{\"kind\": \"READ\", \"substeps\": [\"$1, $2, $3\", \"FROM __stage01_output\"]}, {\"kind\": \"WRITE\", \"substeps\": [\"$1, $2, $3\", \"TO __stage02_output\"]}], \"slotMs\": \"195\"}], \"estimatedBytesProcessed\": \"0\", \"timeline\": [{\"elapsedMs\": \"567\", \"totalSlotMs\": \"4420\", \"pendingUnits\": \"0\", \"completedUnits\": \"52\", \"activeUnits\": \"0\"}, {\"elapsedMs\": \"1195\", \"totalSlotMs\": \"4420\", \"pendingUnits\": \"0\", \"completedUnits\": \"52\", \"activeUnits\": \"0\"}], \"totalPartitionsProcessed\": \"0\", \"totalBytesProcessed\": \"0\", \"totalBytesBilled\": \"0\", \"billingTier\": 0, \"totalSlotMs\": \"4420\", \"cacheHit\": false, \"referencedTables\": [{\"projectId\": \"astronomer-airflow-providers\", \"datasetId\": \"astro_dataset\", \"tableId\": \"table1\"}], \"numDmlAffectedRows\": \"2\", \"statementType\": \"INSERT\", \"dmlStats\": {\"insertedRowCount\": \"2\"}}, \"totalSlotMs\": \"4420\"}, \"status\": {\"state\": \"DONE\"}, \"configuration\": {\"query\": {\"query\": \"INSERT astro_dataset.table1 VALUES (42, 'monthy python', '2022-04-29'), (42, 'fishy fish', '2022-04-29');\", \"destinationTable\": {\"projectId\": \"astronomer-airflow-providers\", \"datasetId\": \"astro_dataset\", \"tableId\": \"table1\"}, \"priority\": \"INTERACTIVE\", \"useLegacySql\": false}, \"jobType\": \"QUERY\"}}"
            },
            "externalQuery": {
                "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.2/integration/airflow",
                "_schemaURL": "https://raw.githubusercontent.com/OpenLineage/OpenLineage/main/spec/OpenLineage.json#/definitions/BaseFacet",
                "externalQueryId": "airflow_example_async_bigquery_queries_insert_query_job_2022_04_29T05_23_10_598413_00_00_22e44b8586a4974727d996d2f1447b33",
                "source": "bigquery"
            }
        },
        "runId": "6c685e0c-4023-4609-bb70-169a248cdd26"
    }
}`


I will close this Spike now and create a separate issue for the implementation as per @phanikumv 's suggestion.
